internal/broker/data.go

```diff
@@ -0,0 +1,58 @@
+package broker
+
+import (
+	"bytes"
+	"encoding/json"
+	"text/template"
+
+	"github.com/ubuntu/oidc-broker/internal/providers/group"
+)
+
+// userInfo represents the user information that is returned to authd.
+type userInfo struct {
+	Name   string
+	UUID   string
+	Home   string
+	Shell  string
+	Gecos  string
+	Groups []group.Info
+}
+
+// MarshalJSON implements the json.Marshaler interface for the userInfo type.
+func (u userInfo) MarshalJSON() ([]byte, error) {
+	var buf bytes.Buffer
+	err := template.Must(template.New("").Parse(`{"userinfo": {
+		"name": "{{.Name}}",
+		"uuid": "{{.UUID}}",
+		"gecos": "{{.Gecos}}",
+		"dir": "{{.Home}}",
+		"shell": "{{.Shell}}",
+		"groups": [ {{range $index, $g := .Groups}}
+			{{- if $index}}, {{end -}}
+			{"name": "{{.Name}}", "ugid": "{{.UGID}}"}
+		{{- end}} ]
+}}`)).Execute(&buf, u)
+	if err != nil {
+		return nil, err
+	}
+
+	return buf.Bytes(), nil
+}
+
+// errorMessage represents the error message that is returned to authd.
+type errorMessage struct {
+	Message string
+}
+
+// MarshalJSON implements the json.Marshaler interface for the errorMessage type.
+func (msg errorMessage) MarshalJSON() ([]byte, error) {
+	if msg.Message == "" {
+		return nil, nil
+	}
+
+	b, err := json.Marshal(msg.Message)
+	if err != nil {
+		return nil, err
+	}
+	return []byte(`{"message": ` + string(b) + `}`), nil
```

Comment:
nitpick: I would still use `fmt.Sprintf()` as explained yesterday for readability

Reply:
Here you are completely ignoring the profiling talks from the sprint...

Reply:
Not sure we are in a performance-sensible place here :)

Reply:
Maintainance > profiling :)

Reply:
To be fair, I don't like too much either the fact we're hardcoding `{"message": ...}` as it is.

As I've a strong allergy with hardcoded code anyways and especially strings, so I'd rather like it more to have `errorMessage` to be fully marshallable. Not sure what was the problem with it though.

Reply:
The issue here is the `json.Marshaler` implementation. We can't call `json.Marshal` directly on the `errorMessage` (otherwise it would recurse infinitely). I feel like the best way here (especially taking the other comment about the userInfo marshaling into consideration), is to define our own interface (something like `dataMarshaler`) to avoid this kind of recursion. Wdyt?

Reply:
Can't we use `json.RawMessage` to stop the recursion?